addons/base_action_rule/models/base_action_rule.py

```diff
@@ -0,0 +1,386 @@
+        if values:
+            model_records.write(values)
+
+        if self.act_followers and hasattr(model_records, 'message_subscribe'):
+            model_records.message_subscribe(self.act_followers.ids)
+
+        # execute server actions
+        if self.server_action_ids:
+            for record in model_records:
+                self.server_action_ids.with_context(active_model=model_records._name, active_ids=[record.id], active_id=record.id).run()
+        return True
+
+    @api.v8
+    def _register_hook(self):
+        """ Patch Models that should trigger action rules based on creation,
+        modification, deletion of records and form onchanges.
+        """
+        #
+        # Note: the patched methods must be defined inside another function,
+        # otherwise their closure may be wrong. For instance, the function
+        # create refers to the outer variable 'create', which you expect to be
+        # bound to create itself. But that expectation is wrong if create is
+        # defined inside a loop; in that case, the variable 'create' is bound to
+        # the last function defined by the loop.
+        #
+
+        def make_create():
+            """ Instanciate a create method that processes action rules. """
+
+            @api.model
+            def create(self, vals):
+                # avoid loops or cascading actions
+                if self.env.context.get('action'):
+                    return create.origin(self, vals)
+
+                # call original method with a modified context
+                created_record = create.origin(self, vals)
+
+                # as it is a new record, we do not consider the actions that have a prefilter
+                # check postconditions, and execute actions on the records that satisfy them
+                for action in self.env['base.action.rule'].with_context(active_test=True).search([('model', '=', self._name),
+                                                                                                  ('kind', 'in', ['on_create', 'on_create_or_write'])]):
+                    if action._filter(action.filter_id, [created_record.id], domain=action.filter_domain):
+                        action._process([created_record.id])
+                return created_record
+            return create
+
+        def make_write():
+            """ Instanciate a _write method that processes action rules. """
+            #
+            # Note: we patch method _write() instead of write() in order to
+            # catch updates made by field recomputations.
+            #
+            @api.multi
+            def _write(self, vals):
+                # avoid loops or cascading actions
+                if self.env.context.get('action'):
+                    return _write.origin(self, vals)
+
+                # retrieve the action rules to possibly execute
+                action_rules = self.env['base.action.rule'].with_context(action=True).search([('model', '=', self._name),
```

Comment:
@vbh-odoo you can pass context with ```with_context``` so as per view and behavior of this its gonna flush the old context which is derived from ```_write``` method so if you do it like : 
```
ctx = dict(self.env.context or {})
ctx['action'] = True
action_rules = self.env['base.action.rule'].with_context(ctx).search(
                                          [
                                                 ('model', '=', self._name),
                                                 ('kind', 'in', ['on_write', 'on_create_or_write'])
                                          ])

```